### PR TITLE
feat: further Enkora importer improvements as agreed with KUVA/Liikunta

### PR DESCRIPTION
**Pull request by Cap-JaTu**
Improvements and changes agreed during Dec -23 and Feb -24

### CODE REVIEW

**Comment by terovirtanen** [comment]
**LINKEDEVENTS-API branch is deployed to platta: https://linkedevents-pr869.api.dev.hel.ninja** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
**LINKEDEVENTS-API branch is deployed to platta: https://linkedevents-pr869.api.dev.hel.ninja** :rocket::rocket::rocket:

**Review by tuomas777 — COMMENTED** [review]
Also, [this docstring](https://github.com/City-of-Helsinki/linkedevents/pull/869/files#diff-a6ba5b39e3e6f6a6d823bbb4a758fded322fa143aa0fd14e5552c98f7a705192R907-R923) seems to be copy-pasted directly from the late Harrastushaku importer, does all of that information really hold true for Enkora importer as well? 🙂

File: events/importer/enkora.py [hunk 1]
```diff
@@ -1019,13 +973,24 @@ def _is_course_expired(course: dict) -> bool:
         )
 
         # Now we have the course list populated, iterate it.
+        if self.options["single"]:
+            logger.info(
+                "Requested sync of a single course: {}".format(self.options["single"])
+            )
+
```
[the remark this answers is not in the document]
Comment (Cap-JaTu) [reply events/importer/enkora.py:980]: It introduces operation mode of doing a single import only. This mode is fully supported by framework and some other importers. Having these "confusing" lines barely enables a framework supported feature.

File: events/importer/enkora.py [hunk 2]
```diff
@@ -1801,37 +1931,89 @@ def convert_location(self, course: dict) -> tuple[Optional[dict], dict]:
...
+        except ObjectDoesNotExist:
+            obj = None
+        if not obj:
+            obj = Place(data_source=self.data_source, origin_id=enkora_place_id)
+            obj._changed = True
+            obj._created = True
+            obj.id = place_id
+            obj._changed_fields = []
+        else:
+            assert obj.id == place_id
```
[the remark this answers is not in the document]
Comment (Cap-JaTu) [reply events/importer/enkora.py:1982]: The check is needed to validate given place existence in database.

Q: What is an improved and project-recommended approach for error handling of missing data?

File: events/importer/enkora.py [hunk 3]
```diff
@@ -1111,8 +1097,138 @@ def mark_deleted(self, event: Event) -> bool:
...
+                course_count, place_count
+            )
+        )
+
+    @staticmethod
+    def _is_course_expired(course: dict, now_is: datetime) -> bool:
+        if "public_visibility_end" not in course:
+            raise ValueError(
+                "Expected to have 'public_visibility_end' field in course! Missing."
+            )
```
[the remark this answers is not in the document]
Comment (Cap-JaTu) [reply events/importer/enkora.py:1213]: Makes sense.

However, doing that will translate into even more changes to code. The optimization target here was to keep number of changes minimal for not to be stuck endlessly in review. This importer is needed badly in production.

File: events/importer/enkora.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (charn) [reply events/importer/enkora.py:1982]: Isn't `obj = Place.objects.get(pk=place_id)` already validating that the place exists in the DB?

**Comment by terovirtanen** [comment]
**LINKEDEVENTS-API branch is deployed to platta: https://linkedevents-pr869.api.dev.hel.ninja** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
**LINKEDEVENTS-API branch is deployed to platta: https://linkedevents-pr869.api.dev.hel.ninja** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
**LINKEDEVENTS-API branch is deployed to platta: https://linkedevents-pr869.api.dev.hel.ninja** :rocket::rocket::rocket:

**Review by tuomas777 — APPROVED** [review]

**Pull request merged**